zkevm-circuits/src/evm_circuit/execution/signed_comparator.rs

```diff
@@ -0,0 +1,267 @@
+    ) -> Result<(), Error> {
+        self.same_context.assign_exec_step(region, offset, step)?;
+
+        let opcode = step.opcode.unwrap();
+
+        // SLT opcode is the default check in the SCMP gadget. Swap rw for SGT.
+        let is_sgt = self.is_sgt.assign(
+            region,
+            offset,
+            F::from(opcode.as_u8() as u64),
+            F::from(OpcodeId::SGT.as_u8() as u64),
+        )?;
+        let indices = if is_sgt == F::one() {
+            [step.rw_indices[1], step.rw_indices[0]]
+        } else {
+            [step.rw_indices[0], step.rw_indices[1]]
+        };
+        let [a, b] =
+            indices.map(|idx| block.rws[idx].stack_value().to_le_bytes());
+
+        self.sign_check_a.assign(
+            region,
+            offset,
+            F::from(a[0] as u64),
+            F::from(128u64),
+        )?;
+        self.sign_check_b.assign(
+            region,
+            offset,
+            F::from(b[0] as u64),
+            F::from(128u64),
+        )?;
+
+        self.lt_lo.assign(
+            region,
+            offset,
+            from_bytes::value(&a[0..16]),
+            from_bytes::value(&b[0..16]),
+        )?;
+        self.lt_hi.assign(
+            region,
+            offset,
+            from_bytes::value(&a[16..32]),
+            from_bytes::value(&b[16..32]),
+        )?;
+
+        self.a.assign(region, offset, Some(a))?;
+        self.b.assign(region, offset, Some(b))?;
+
+        Ok(())
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use bus_mapping::{bytecode, eth_types::Word, evm::OpcodeId};
+
+    use crate::evm_circuit::{
+        test::{rand_word, run_test_circuit_incomplete_fixed_table},
+        witness,
+    };
+
+    use std::convert::TryInto;
+
+    fn test_ok(opcode: OpcodeId, a: Word, b: Word) {
+        let bytecode = bytecode! {
+            PUSH32(b)
+            PUSH32(a)
+            #[start]
+            .write_op(opcode)
+            STOP
+        };
+        let block = witness::build_block_from_trace_code_at_start(&bytecode);
+        assert_eq!(run_test_circuit_incomplete_fixed_table(block), Ok(()));
+    }
+
+    #[test]
+    fn signed_comparator_gadget_simple() {
+        let zero = Word::from_big_endian(&[0u8; 32]);
+        let minus_1 = Word::from_big_endian(&[255u8; 16]);
+        let minus_2 = {
+            let mut bytes = vec![255u8; 15];
+            bytes.push(254u8);
+            let bytes: [u8; 16] = bytes.try_into().unwrap();
+            Word::from_big_endian(&bytes)
+        };
+        let plus_1 = {
+            let mut bytes = vec![0u8; 15];
+            bytes.push(1);
+            let bytes: [u8; 16] = bytes.try_into().unwrap();
+            Word::from_big_endian(&bytes)
+        };
+
+        // SLT
+        test_ok(OpcodeId::SLT, minus_1, zero);
+        test_ok(OpcodeId::SLT, zero, plus_1);
+        test_ok(OpcodeId::SLT, minus_1, zero);
+        test_ok(OpcodeId::SLT, minus_2, plus_1);
+        test_ok(OpcodeId::SLT, minus_2, minus_1);
```

Comment:
add one extra test where `a_hi > b_hi` and `a_lo < b_lo` when both `a` and `b` are positive